src/transformers/trainer.py

```diff
@@ -847,6 +847,17 @@ def create_optimizer(self):
                 )
             else:
                 self.optimizer = optimizer_cls(optimizer_grouped_parameters, **optimizer_kwargs)
+                if optimizer_cls.__name__ == "Adam8bit":
+                    from torch.nn import Embedding
+
+                    import bitsandbytes
+
+                    manager = bitsandbytes.optim.GlobalOptimManager.get_instance()
+
+                    for module in self.model.modules():
+                        if isinstance(module, Embedding):
+                            manager.register_module_override(module, "weight", {"optim_bits": 32})
+                            logger.info(f"Registering bitsandbytes override for {module}")
```

Comment:
This can all go in `get_optimizer_cls_and_kwargs` if we make it a real method. There is no real reason for it to be a static method anyway (unless I'm missing something @stas00 ?). It would be cleaner to have all the optimizer-specific logic in that method.

Reply:
Keeping all the optimizer-specific code together would be cleaner (especially given that in `create_optimizer`, I check the optimizer by string class name). A regular method would help with that.

However, we might also need to override the `bnb` configuration before the model is moved to the GPU (see my initial comment). I hope @TimDettmers can clarify that for us.

Reply:
Yes, let's hear from Tim and then try to follow Sylvain's recommendation if it works. Otherwise keep what you did.